src/mbgl/map/map_context.cpp

```diff
@@ -115,15 +125,110 @@ void MapContext::updateTiles() {
+            auto& shapeStyle = annotation->styleProperties;
+
+            // apply shape paint properties
+            ClassProperties paintProperties;
+
+            if (shapeStyle.is<LineProperties>()) {
+                // opacity
+                PropertyValue lineOpacity = ConstantFunction<float>(shapeStyle.get<LineProperties>().opacity);
+                paintProperties.set(PropertyKey::LineOpacity, lineOpacity);
+
+                // line width
+                PropertyValue lineWidth = ConstantFunction<float>(shapeStyle.get<LineProperties>().width);
+                paintProperties.set(PropertyKey::LineWidth, lineWidth);
+
+                // stroke color
+                PropertyValue strokeColor = ConstantFunction<Color>(shapeStyle.get<LineProperties>().color);
+                paintProperties.set(PropertyKey::LineColor, strokeColor);
+            } else if (shapeStyle.is<FillProperties>()) {
+                // opacity
+                PropertyValue fillOpacity = ConstantFunction<float>(shapeStyle.get<FillProperties>().opacity);
+                paintProperties.set(PropertyKey::FillOpacity, fillOpacity);
+
+                // fill color
+                PropertyValue fillColor = ConstantFunction<Color>(shapeStyle.get<FillProperties>().fill_color);
+                paintProperties.set(PropertyKey::FillColor, fillColor);
+
+                // stroke color
+                PropertyValue strokeColor = ConstantFunction<Color>(shapeStyle.get<FillProperties>().stroke_color);
+                paintProperties.set(PropertyKey::FillOutlineColor, strokeColor);
+            }
+
+            std::map<ClassID, ClassProperties> shapePaints;
+            shapePaints.emplace(ClassID::Default, std::move(paintProperties));
+
+            // create shape layer
+            util::ptr<StyleLayer> shapeLayer = std::make_shared<StyleLayer>(shapeLayerID, std::move(shapePaints));
+            shapeLayer->type = (shapeStyle.is<LineProperties>() ? StyleLayerType::Line : StyleLayerType::Fill);
+            style->layers.emplace_back(shapeLayer);
+
+            // create shape bucket & connect to source
+            util::ptr<StyleBucket> shapeBucket = std::make_shared<StyleBucket>(shapeLayer->type);
+            shapeBucket->name = shapeLayer->id;
+            shapeBucket->source_layer = shapeLayer->id;
+            shapeBucket->source = *source_it;
+
+            // apply line layout properties to bucket
+            if (shapeStyle.is<LineProperties>()) {
+                shapeBucket->layout.set(PropertyKey::LineJoin, ConstantFunction<JoinType>(JoinType::Round));
+            }
+
+            // connect layer to bucket
+            shapeLayer->bucket = shapeBucket;
+        }
+    }
+
+    // invalidate annotations layer tiles
     for (const auto &source : style->sources) {
         if (source->info.type == SourceType::Annotations) {
             source->invalidateTiles(ids);
         }
     }
-    triggerUpdate();
+
+    cascadeClasses();
```

Comment:
Why does this both call `cascadeClasses` and `triggerUpdate(Update::Classes)`? The latter should be sufficient.